composer/models/bert/model.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2021 MosaicML. All Rights Reserved.
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Mapping, Tuple
+
+import transformers
+from torchmetrics import Accuracy, MatthewsCorrcoef, MeanSquaredError, SpearmanCorrcoef
+from torchmetrics.collections import MetricCollection
+
+from composer.models.nlp_metrics import BinaryF1Score, CrossEntropyLoss, MaskedAccuracy
+from composer.models.transformer_shared import MosaicTransformer
+
+if TYPE_CHECKING:
+    from composer.core.types import Batch, BatchDict, Metrics, Tensors
+
+
+class BERTModel(MosaicTransformer):
+    """
+    Implements a BERT wrapper around a MosaicTransformer.
+    """
+
+    def __init__(self, module: transformers.BertModel, config: transformers.BertConfig, tokenizer_name: str) -> None:
+        super().__init__(
+            module=module,  #type: ignore (thirdparty)
+            config=config,
+            tokenizer_name=tokenizer_name)
+
+        # we're going to remove the label from the expected inputs
+        # since we will handle metric calculation with TorchMetrics instead of HuggingFace.
+        self.model_inputs.remove("labels")
+
+        self.train_metrics = []
+        self.val_metrics = []
+
+        # TODO (Moin): make sure this is moved to be dataset-specific
+        # if config.num_labels=1, then we are training a regression task, so we should update our loss functions
```

Comment:
> I think there should probably be something like config.task_type: ["regression", "binary", "classification", "mlm"] rather than just a check on config.num_labels. Then different workloads can specify in the BERTConfig what type of model head+loss they want to use.

It's important to note that the current task-specific metric calculation is just a temporary stopgap until the "evaluation on multiple datasets" PR is merged. At that point, the metrics and losses will be a function of the dataset, not the model. This will naturally give us what you are asking for, and I agree with your statement. I just implemented something quickly so when the multiple datasets PR is ready, we can switch over.

cc @Landanjs I think you had the same comment

Reply:
Ack that the future Evaluator framework solves the metrics and loss functions. And I guess we won't have switches for heads within `BERTModel`, but instead multiple higher level classes like `BERTForClassification`?